### ISSUE TITLE
`rustc` v1.64.0 hangs for a crate with a syntax error

### ISSUE THREAD

**Issue by tdyas** [rust-lang/rust#103823]
Issue: `rustc` v1.64.0 hangs on a crate with a syntax error without reporting the error. 

I was making [this change](https://github.com/pantsbuild/pants/pull/17290) to a [Rust project](https://github.com/pantsbuild/pants) where one of the non-default workspace crates ended up with an error due to the signature of a function in another crate in the workspace changing. Normally, this crate compiles relatively quickly, but after making the change, `rustc` ran indefinitely instead of showing the error. Fixing the error (or not making the change in the first place) allows the crate to compile normally again.

Seems to build fine on nightly as of today.

Commits:
- [Commit which triggers the hang](https://github.com/pantsbuild/pants/pull/17290/commits/0f7dd58f884300740aa9f0e2fd96637b8fe2616e)
- [Commit which fixes the syntax error and then crate compiles normally](https://github.com/pantsbuild/pants/pull/17290/commits/7c0024f76b990f8513514f65f52127f53d005bf2)

Reproduces on both Linux and macOS/ARM with Rust v1.64.0. Nightly appears to build without a hang and then displays the expected syntax error.

How to reproduce:

1. Clone [my fork](https://github.com/tdyas/pants) of the https://github.com/pantsbuild/pants/ repository.
2. Checkout the commit: `git checkout 0f7dd58f884300740aa9f0e2fd96637b8fe2616e`
3. Build the crate at issue:
  ```
  cd src/rust/engine/process_executor
  cargo build
  ```
4. The build should hang. Reproduces on both Linux and macOS/ARM with Rust v1.64.0.

Traces:
- [Thread dumps #1 and #2](https://github.com/rust-lang/rust/files/9905760/thread-dumps.txt)
- [perf.hist.txt](https://github.com/rust-lang/rust/files/9905753/perf.hist.txt)

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: aarch64-apple-darwin
release: 1.64.0
LLVM version: 14.0.6
```

Nightly version info on which there is no hang:

```
rustc 1.67.0-nightly (77e57db38 2022-10-30)
binary: rustc
commit-hash: 77e57db384aca99444c3b5f6a9c86bc58a804d89
commit-date: 2022-10-30
host: aarch64-apple-darwin
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by saethlin** [comment]
Based on the profile, this loop seems relevant and also suspicious, but I'm miles out of my depth: https://github.com/rust-lang/rust/blob/0940040c0486a536be4f8685c7dd9a078f9e87c2/compiler/rustc_hir_typeck/src/fn_ctxt/arg_matrix.rs#L228

**Comment by compiler-errors** [comment]
Since it's not hanging on nightly, this probably can't be fixed, since Rust 1.65 comes out in a just few days.

**Comment by chenyukang** [comment]
It's same issue with #100478, fixed by https://github.com/rust-lang/rust/pull/100502
